scripts/mayachemtools/docs/modules/man3/FileIO.3
.\" Automatically generated by Pod::Man 2.28 (Pod::Simple 3.35)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "FileIO 1"
.TH FileIO 1 "2018-05-15" "perl v5.22.4" "MayaChemTools"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
FileIO
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
use FileIO::FileIO;
.PP
use FileIO::FileIO qw(:all);
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBFIleIO\fR class provides following methods:
.PP
new, Close, Open, SetMode
.PP
\&\fBFleIO\fR class serves as a base class for all classes involved in file \s-1IO.\s0 It is derived from
\&\fBObjectProperty\fR base class which provides methods not explicitly defined in \fBAtom\fR or
\&\fBObjectProperty\fR class using Perl's \s-1AUTOLOAD\s0 functionality. These methods
are generated on-the-fly for a specified object property:
.PP
.Vb 3
\&    Set<PropertyName>(<PropertyValue>);
\&    $PropertyValue = Get<PropertyName>();
\&    Delete<PropertyName>();
.Ve
.SS "\s-1METHODS\s0"
.IX Subsection "METHODS"
.IP "\fBnew\fR" 4
.IX Item "new"
.Vb 1
\&    $NewFileIO = new FileIO(%NamesAndValues);
.Ve
.Sp
Using specified \fIFileIO\fR property names and values hash, \fBnew\fR method creates a new object
and returns a reference to a newly created \fBFileIO\fR object. By default, the following properties are
initialized:
.Sp
.Vb 4
\&    Name = \*(Aq\*(Aq;
\&    Mode = \*(AqRead\*(Aq;
\&    Status = 0;
\&    FileHandle = \*(Aq\*(Aq;
.Ve
.IP "\fBClose\fR" 4
.IX Item "Close"
.Vb 1
\&    $FileIO\->Close();
.Ve
.Sp
Close open file and returns \fIFileIO\fR.
.IP "\fBOpen\fR" 4
.IX Item "Open"
.Vb 1
\&    $FileIO\->Open();
.Ve
.Sp
Opens the file using file \fIName\fR and \fIMode\fR and returns \fIFileIO\fR.
.IP "\fBSetMode\fR" 4
.IX Item "SetMode"
.Vb 1
\&    $FileIO\->SetMode($Mode);
.Ve
.Sp
Sets up file \fIMode\fR and returns \fIFileIO\fR Default \fIMode\fR value: \fIRead\fR.
Supported \fIMode\fR values:
.Sp
.Vb 1
\&    Read, Write, Append, <, >, >>, r, w, a
.Ve
.IP "\fBSetName\fR" 4
.IX Item "SetName"
.Vb 1
\&    $FileIO\->SetName($Name);
.Ve
.Sp
Sets up file \fIName\fR and returns \fIFileIO\fR.
.SH "AUTHOR"
.IX Header "AUTHOR"
Manish Sud <msud@san.rr.com>
.SH "SEE ALSO"
.IX Header "SEE ALSO"
MoleculeFileIO.pm, MDLMolFileIO.pm, SDFileIO.pm
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (C) 2018 Manish Sud. All rights reserved.
.PP
This file is part of MayaChemTools.
.PP
MayaChemTools is free software; you can redistribute it and/or modify it under
the terms of the \s-1GNU\s0 Lesser General Public License as published by the Free
Software Foundation; either version 3 of the License, or (at your option)
any later version.


mayachemtools/lib/FileIO/FileIO.pm
package FileIO::FileIO;
#
# File: FileIO.pm
# Author: Manish Sud <msud@san.rr.com>
#
# Copyright (C) 2018 Manish Sud. All rights reserved.
#
# This file is part of MayaChemTools.
#
# MayaChemTools is free software; you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by the Free
# Software Foundation; either version 3 of the License, or (at your option) any
# later version.
#
# MayaChemTools is distributed in the hope that it will be useful, but without
# any warranty; without even the implied warranty of merchantability of fitness
# for a particular purpose.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with MayaChemTools; if not, see <http://www.gnu.org/licenses/> or
# write to the Free Software Foundation Inc., 59 Temple Place, Suite 330,
# Boston, MA, 02111-1307, USA.
#

use strict;
use Carp;
use Exporter;
use FileHandle;
use ObjectProperty;

use vars qw(@ISA @EXPORT @EXPORT_OK %EXPORT_TAGS);

@ISA = qw(ObjectProperty Exporter);
@EXPORT = qw();
@EXPORT_OK = qw();

%EXPORT_TAGS = (all  => [@EXPORT, @EXPORT_OK]);

# Setup class variables...
my($ClassName);
_InitializeClass();

# Class constructor...
sub new {
  my($Class, %NamesAndValues) = @_;

  # Initialize object...
  my $This = {};
  bless $This, ref($Class) || $Class;
  $This->_InitializeFileIO();

  $This->_InitializeFileIOProperties(%NamesAndValues);

  return $This;
}

# Initialize object data...
#
sub _InitializeFileIO {
  my($This) = @_;

  # File name...
  $This->{Name} = '';

  # Read, write or append...
  $This->{Mode} = 'Read';

  # Open/close status...
  $This->{Status} = 0;

  # File handle returned by file open...
  $This->{FileHandle} = '';
}

# Initialize class ...
sub _InitializeClass {
  #Class name...
  $ClassName = __PACKAGE__;

}

# Initialize object properties....
sub _InitializeFileIOProperties {
  my($This, %NamesAndValues) = @_;

  my($Name, $Value, $MethodName);
  while (($Name, $Value) = each  %NamesAndValues) {
    $MethodName = "Set${Name}";
    $This->$MethodName($Value);
  }

  return $This;
}

# Close any open file...
sub DESTROY {
  my($This) = @_;

  $This->Close();

  return $This;
}

# Set file name and make sure it's not already set...
#
sub SetName {
  my($This, $Name) = @_;

  if ($This->{Name}) {
    croak "Error: ${ClassName}->SetName: Can't set file name to $Name:  $This->{Name}...";
  }

  $This->{Name} = $Name;

  return $This;
}

# Open file using specified mode...
#
sub Open {
  my($This, $Mode) = @_;

  if ($This->{Status}) {
    croak "Error: ${ClassName}->Open: Can't open file $This->{Name}: It's already open...";
  }

  if (defined $Mode) {
    # Set mode...
    $This->SetMode($Mode);
  }

  # Get name and mode...
  my($Name);
  $Name = $This->{Name};
  $Mode = $This->_GetOpenMode();

  # Open the file using specified mode and store FileHandle...
  my($FileHandle);
  $FileHandle = new FileHandle("${Mode}${Name}");
  if (!defined $FileHandle) {
    croak "Error: ${ClassName}->Open: Can't open $Name:  $! ...";
  }
  $This->{FileHandle} = $FileHandle;
  $This->{Status} = 1;

  return $This;
}

# Close an open file...
sub Close {
  my($This) = @_;

  if ($This->{Status}) {
    $This->{FileHandle}->close();
  }
  $This->{Status} = 0;

  return $This;
}

# Supported Mode values are: Read, Write, Append, <, >, >>, r, w, a
#
sub SetMode {
  my($This, $SpecifiedMode) = @_;
  my($Mode);

  if (!defined $SpecifiedMode) {
    $SpecifiedMode = 'Read';
  }

  MODE: {
    if ($SpecifiedMode =~ /^(Read|<|r)$/i) { $Mode = 'Read'; last MODE; }
    if ($SpecifiedMode =~ /^(Write|>|w)$/i) { $Mode = 'Write'; last MODE; }
    if ($SpecifiedMode =~ /^(Append|>>|a)$/i) { $Mode = 'Append'; last MODE; }
    croak "Error: ${ClassName}->SetMode: Specified mode value, $SpecifiedMode, is not valid: Supported values: Read, Write, Append, <, >, >>, r, w, a...";
  }
  $This->{Mode} = $Mode;

  return $This;
}

# Get mode values to be used for file open function: <, >, >>
#
sub _GetOpenMode {
  my($This) = @_;
  my($Mode);

  MODE: {
    if ($This->{Mode} =~ /^(Read|<|r)$/i) { $Mode = '<'; last MODE; }
    if ($This->{Mode} =~ /^(Write|>|w)$/i) { $Mode = '>'; last MODE; }
    if ($This->{Mode} =~ /^(Append|>>|a)$/i) { $Mode = '>>'; last MODE; }
    $Mode = '';
  }
  return $Mode;
}

1;

__END__

=head1 NAME

FileIO

=head1 SYNOPSIS

use FileIO::FileIO;

use FileIO::FileIO qw(:all);

=head1 DESCRIPTION

B<FIleIO> class provides following methods:

new, Close, Open, SetMode

B<FleIO> class serves as a base class for all classes involved in file IO. It is derived from
B<ObjectProperty> base class which provides methods not explicitly defined in B<Atom> or
B<ObjectProperty> class using Perl's AUTOLOAD functionality. These methods
are generated on-the-fly for a specified object property:

    Set<PropertyName>(<PropertyValue>);
    $PropertyValue = Get<PropertyName>();
    Delete<PropertyName>();

=head2 METHODS

=over 4

=item B<new>

    $NewFileIO = new FileIO(%NamesAndValues);

Using specified I<FileIO> property names and values hash, B<new> method creates a new object
and returns a reference to a newly created B<FileIO> object. By default, the following properties are
initialized:

    Name = '';
    Mode = 'Read';
    Status = 0;
    FileHandle = '';

=item B<Close>

    $FileIO->Close();

Close open file and returns I<FileIO>.

=item B<Open>

    $FileIO->Open();

Opens the file using file I<Name> and I<Mode> and returns I<FileIO>.

=item B<SetMode>

    $FileIO->SetMode($Mode);

Sets up file I<Mode> and returns I<FileIO> Default I<Mode> value: I<Read>.
Supported I<Mode> values:

    Read, Write, Append, <, >, >>, r, w, a

=item B<SetName>

    $FileIO->SetName($Name);

Sets up file I<Name> and returns I<FileIO>.

=back

=head1 AUTHOR

Manish Sud <msud@san.rr.com>

=head1 SEE ALSO

MoleculeFileIO.pm, MDLMolFileIO.pm, SDFileIO.pm

=head1 COPYRIGHT

Copyright (C) 2018 Manish Sud. All rights reserved.

This file is part of MayaChemTools.

MayaChemTools is free software; you can redistribute it and/or modify it under
the terms of the GNU Lesser General Public License as published by the Free
Software Foundation; either version 3 of the License, or (at your option)
any later version.

=cut
